### PR TITLE
fixes for clang -Os on clang >= 12.0.0

### PR DESCRIPTION
**Pull request by douzzer**
fixes for clang -Os on clang >= 12.0.0 (aggressive layout and reordering optimizations);

fixes for bugs in blake2s.

### CODE REVIEW

File: wolfcrypt/src/sp_int.c [hunk 1]
```diff
@@ -93,6 +93,11 @@ This library provides single precision (SP) integer math functions.
  * WOLFSSL_SP_FAST_MODEXP       Allow fast mod_exp with small C code
  */
 
+/* WOLFSSL_SP_SMALL is incompatibile with clang -Os. */
+#if defined(__clang__) && defined(WOLFSSL_SP_SMALL)
```
Comment (SparkiDev) [inline wolfcrypt/src/sp_int.c:97]: Please check version 12 here.

File: wolfcrypt/src/blake2s.c [hunk 2]
```diff
@@ -132,24 +136,27 @@ int blake2s_init( blake2s_state *S, const byte outlen )
   store32( &P->node_offset, 0 );
   P->node_depth    = 0;
   P->inner_length  = 0;
-  XMEMSET( P->reserved, 0, sizeof( P->reserved ) );
   XMEMSET( P->salt,     0, sizeof( P->salt ) );
   XMEMSET( P->personal, 0, sizeof( P->personal ) );
 #else
-  XMEMSET( P, 0, sizeof( *P ) );
+  XMEMSET( (blake2s_param *)P, 0, sizeof( *P ) );
```
Comment (SparkiDev) [inline wolfcrypt/src/blake2s.c:142]: Why is the cast necessary?

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: wolfcrypt/src/blake2s.c (hunk 2 above)
Comment (douzzer) [reply wolfcrypt/src/blake2s.c:142]: compiler warning otherwise, around the `volatile` qualifier.

File: wolfcrypt/src/sp_int.c (hunk 1 above)
Comment (douzzer) [reply wolfcrypt/src/sp_int.c]: done

File: wolfcrypt/src/sp_int.c [hunk 3]
```diff
@@ -93,6 +93,12 @@ This library provides single precision (SP) integer math functions.
  * WOLFSSL_SP_FAST_MODEXP       Allow fast mod_exp with small C code
  */
 
+/* WOLFSSL_SP_SMALL is incompatibile with clang-12+ -Os. */
```
Comment (dgarske) [inline wolfcrypt/src/sp_int.c:96]: Spell `incompatible`. Also can we flag this with a `TODO:`. I'm worried the issues here will be overlooked later.

File: wolfssl/wolfcrypt/types.h [hunk 4]
```diff
@@ -761,13 +761,12 @@ decouple library dependencies with standard string, memory and so on.
         #define XSTRLCAT(s1,s2,n) wc_strlcat((s1),(s2),(n))
     #endif
 
-    #if !defined(NO_FILESYSTEM) && defined(OPENSSL_EXTRA) && \
-        !defined(NO_STDIO_FILESYSTEM)
+    #if !defined(NO_FILESYSTEM) && !defined(NO_STDIO_FILESYSTEM)
```
Comment (dgarske) [inline wolfssl/wolfcrypt/types.h:764]: Worried about the stdlib.h for embedded customers. Can you include `HAVE_STDLIB_H` or do you think `NO_STDIO_FILESYSTEM` will be good enough?

File: wolfssl/wolfcrypt/blake2-int.h [hunk 5]
```diff
@@ -73,7 +73,7 @@
     byte  node_depth;    /* 15 */
     byte  inner_length;  /* 16 */
     /* byte  reserved[0]; */
-    byte  salt[BLAKE2B_SALTBYTES]; /* 24 */
+    byte  salt[BLAKE2S_SALTBYTES]; /* 24 */
```
Comment (dgarske) [inline wolfssl/wolfcrypt/blake2-int.h:76]: Thanks for fixing the copy/paste error here... which used the wrong salt size for blake2b.... 16 vs 24.

File: testsuite/testsuite.c [hunk 6]
```diff
@@ -241,7 +256,7 @@ static int test_tls(func_args* server_args)
 {
     func_args echo_args;
     char* myArgv[NUMARGS];
-    char arg[3][32];
+    char arg[3][128];
```
Comment (dgarske) [inline testsuite/testsuite.c:259]: Curious why the arg size increase?

File: testsuite/testsuite.c [hunk 7]
```diff
@@ -104,19 +104,34 @@ int testsuite_test(int argc, char** argv)
...
+    const int tempName_len = 8;
+    const int tempName_Xnum = 6;
 #endif
 #ifdef HAVE_STACK_SIZE
     void *serverThreadStackContext = NULL;
 #endif
     int ret;
 
+#ifndef USE_WINDOWS_API
+#ifdef XGETENV
```
Comment (dgarske) [inline testsuite/testsuite.c:122]: Great work here!

File: fips-check.sh [hunk 8]
```diff
@@ -184,6 +184,7 @@ linuxv2)
   FIPS_SRCS+=( wolfcrypt_first.c wolfcrypt_last.c )
   FIPS_INCS=( fips.h )
   FIPS_OPTION=v2
+  COPY_DIRECT=( wolfcrypt/src/aes_asm.S wolfcrypt/src/aes_asm.asm )
```
Comment (dgarske) [inline fips-check.sh:187]: Thanks for fixing the aes_asm file for FIPS. Thumbs up

**Review by dgarske — CHANGES_REQUESTED** [review]
This PR is awesome. Just a few minor comments.

File: testsuite/testsuite.c (hunk 6 above)
Comment (douzzer) [reply testsuite/testsuite.c:259]: to accommodate `$TMPDIR`.  ran into the failure mode empirically :-)

File: wolfssl/wolfcrypt/types.h (hunk 4 above)
Comment (douzzer) [reply wolfssl/wolfcrypt/types.h:764]: I did some more research on this.  we have a slew of other `#include <stdlib.h>`s around and about, gated on various things like particular ports, `!defined(XATOI)`, `defined(WOLFSSL_MEMORY_LOG)`, `defined(NO_WOLFSSL_MEMORY) && !defined(WOLFSSL_NO_MALLOC)`, and `!defined(STRING_USER) && !defined(WOLFSSL_LINUXKM)`.

we're not using `HAVE_STDLIB_H` anywhere in the code -- not sure it's even getting detected by or exported from autotools.

**Review by dgarske — APPROVED** [review]

**Pull request merged**